src/signer/canister/src/sign/bitcoin/bitcoin_api.rs

```diff
@@ -18,3 +20,48 @@ pub async fn get_balance(network: BitcoinNetwork, address: String) -> Result<u64
 
     Ok(balance_res.0)
 }
+
+/// Returns the 100 fee percentiles measured in millisatoshi/byte.
+/// Percentiles are computed from the last 10,000 transactions (if available).
+///
+/// Relies on the `bitcoin_get_current_fee_percentiles` endpoint.
+/// See [Bitcoin API](https://internetcomputer.org/docs/current/references/ic-interface-spec/#ic-bitcoin_get_current_fee_percentiles)
+async fn get_current_fee_percentiles(
+    network: BitcoinNetwork,
+) -> Result<Vec<MillisatoshiPerByte>, String> {
+    let res = bitcoin_get_current_fee_percentiles(GetCurrentFeePercentilesRequest { network })
+        .await
+        .map_err(|err| err.1)?;
+
+    Ok(res.0)
+}
+
+/// Returns the 50th percentile for sending fees.
+pub async fn get_fee_per_byte(network: BitcoinNetwork) -> Result<u64, String> {
+    // Get fee percentiles from previous transactions to estimate our own fee.
+    let fee_percentiles = get_current_fee_percentiles(network).await?;
+
+    if fee_percentiles.is_empty() {
+        // There are no fee percentiles. This case can only happen on a regtest
+        // network where there are no non-coinbase transactions. In this case,
+        // we use a default of 2000 millisatoshis/byte (i.e. 2 satoshi/byte)
+        Ok(2000)
+    } else {
+        Ok(fee_percentiles[50])
+    }
```

Comment:
How do you know that the 50th cell is there?  The docs suggest implicitly that the vector contains percentages, so presumably has 101 cells (0-100 inclusive) but this doesn't seem to be guaranteed in the API or code.  Is it possible for the vector to be shorter?  It would be nice if the ic_cdk rust API were to guarantee the length (if the length is guaranteed).

Maybe that could be handled in the same way as when there are no percentiles:

```suggestion
        // There may be no fee percentiles. This case can only happen on a regtest
        // network where there are no non-coinbase transactions. In this case,
        // we use a default of 2000 millisatoshis/byte (i.e. 2 satoshi/byte)
   Ok(fee_percentiles.get(50).unwrap_or(2000))
```

Reply:
Nice!

I also asked the team on slack: https://dfinity.slack.com/archives/CGZJ7G1J6/p1727249531022879

But it makes sense to use the length to get the middle point.